### PR TITLE
Add a test for gc dropping active layers (#707)

### PR DESCRIPTION
**Pull request by dhammika**
Signed-off-by: Dhammika Pathirana <dhammika@gmail.com>

### CODE REVIEW

**Comment by dhammika** [comment]
Test also relies on low compaction_threshold to trigger layer compaction but that's not configurable atm.   
```diff
diff --git a/pageserver/src/layered_repository.rs b/pageserver/src/layered_repository.rs
index d7a250f3..3fc04a5c 100644
--- a/pageserver/src/layered_repository.rs
+++ b/pageserver/src/layered_repository.rs
@@ -1683,7 +1683,7 @@ impl LayeredTimeline {
 
         // We compact or "shuffle" the level-0 delta layers when 10 have
         // accumulated.
-        static COMPACT_THRESHOLD: usize = 10;
+        static COMPACT_THRESHOLD: usize = 2;
 
         let level0_deltas = layers.get_level0_deltas()?;
 
```

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -1681,13 +1681,9 @@ impl LayeredTimeline {
...
         let layers = self.layers.lock().unwrap();
 
-        // We compact or "shuffle" the level-0 delta layers when 10 have
-        // accumulated.
-        static COMPACT_THRESHOLD: usize = 10;
-
         let level0_deltas = layers.get_level0_deltas()?;
 
-        if level0_deltas.len() < COMPACT_THRESHOLD {
+        if level0_deltas.len() < self.conf.compaction_threshold {
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1686]: Let's keep the old comment, "We compact...." here

**Review by hlinnaka — APPROVED** [review]
The first and second commit ought to be squashed together, as the first commit depends on the second one. Those changes look OK to me, let's squash & push them.

If I understand correctly, the third commit, "Add a test for gc dropping active layers (#707)", doesn't actually reproduce the problem. But if you set `compaction_threshold=2` in the test, then it does? Let's hold off on the third commit, and push it together with the fix for the issue, when we have a fix.

One idea for testing this: We could add a new internal "compact" command to the pageserver API, similar to the "do_gc" and "checkpoint" commands, to force immediate compaction. You could call that in the test to force compaction at just the right time, to reproduce the #707 scenario.

**Comment by dhammika** [comment]
> The first and second commit ought to be squashed together, as the first commit depends on the second one. Those changes look OK to me, let's squash & push them.

Merge test script patches, https://github.com/neondatabase/neon/pull/1484/commits/bbfed6216e8ef0f66e7af343bcfb1004533923b6 and https://github.com/neondatabase/neon/pull/1484/commits/f489d64cfa3854a70358d1d6cca539cfa0c368a9?

> 
> If I understand correctly, the third commit, "Add a test for gc dropping active layers (#707)", doesn't actually reproduce the problem. But if you set `compaction_threshold=2` in the test, then it does? Let's hold off on the third commit, and push it together with the fix for the issue, when we have a fix.

Ya, get a page load error, 
```
IoError: could not read block 19 in rel 1663/12974/2691.0 from page server at lsn 0/0...
```

> 
> One idea for testing this: We could add a new internal "compact" command to the pageserver API, similar to the "do_gc" and "checkpoint" commands, to force immediate compaction. You could call that in the test to force compaction at just the right time, to reproduce the #707 scenario.

Sure, I'll check this.

File: pageserver/src/layered_repository.rs (hunk 1 above)
Comment (dhammika) [reply pageserver/src/layered_repository.rs:1686]: Moved config patch to #1497

**Comment by dhammika** [comment]
Looks like the child doesn't traverse ancestor's layers after gc run. It's terminating the key search after inmem layer, `cont_lsn 0/16CAC19, layer: inmem-00000000016CAC19-FFFFFFFFFFFFFFFF`

```
2022-04-18T19:36:00.424625Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T19:36:00.424674Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CHECKING for 000000000000000000000000000000000000 at 0/16D12B9 on open layer inmem-00000000016CAC19-FFFFFFFFFFFFFFFF
2022-04-18T19:36:00.424723Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T19:36:00.424760Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T19:36:00.424826Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T19:36:00.424889Z ERROR WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: PATH: result Continue, cont_lsn 0/16CAC19, layer: inmem-00000000016CAC19-FFFFFFFFFFFFFFFF
```

Trace of the LSN=`0/16D12B8` without gc shows ancestor's layer traversal. 
```
2022-04-18T20:00:14.527410Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T20:00:14.527483Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CHECKING for 000000000000000000000000000000000000 at 0/16D12B9 on open layer inmem-00000000016CAC19-FFFFFFFFFFFFFFFF
2022-04-18T20:00:14.527559Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T20:00:14.527623Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T20:00:14.527709Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CHECKING for 000000000000000000000000000000000000 at 0/16CAC19 on historic layer 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001696629-0000000001AA3581
2022-04-18T20:00:14.527811Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T20:00:14.527877Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/0: Continue with 0 records
2022-04-18T20:00:14.527960Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CHECKING for 000000000000000000000000000000000000 at 0/1696629 on historic layer 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016965B0-0000000001696629
2022-04-18T20:00:14.528107Z  INFO WAL receiver{timeline=19fa58bcf78055cca1dc94e4e945c09a tenant=63fdd25b3949405e928b9877c9d4a5d4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}: CALLED for 0/16D12B8 at 0/16965B0: Complete with 0 records
```

Child timeline's ancestor_lsn shows `0/1696628` though it was branched at `0/16CAC18`
```
2022-04-18T19:36:00.425441Z  INFO WAL receiver{timeline=33a103820161957f4530e6f97da1aa3d tenant=a43e322b29e642f28e7d77791d2791d5}: WAL streaming connection failed (could not find layer with more data for key 000000000000000000000000000000000000 at LSN 0/16CAC18, request LSN 0/16D12B8, ancestor 0/1696628)
```
and branch point, 
```
2022-04-18T19:35:46.453330Z  INFO /timeline_create{tenant=a43e322b29e642f28e7d77791d2791d5 new_timeline=None lsn=Some(0/16CAC18)}: branched timeline 33a103820161957f4530e6f97da1aa3d from e2dcbc351a945db58ed8902fdcd0fda4 at 0/16CAC18
```

**Comment by dhammika** [comment]
```
2022-04-19T22:26:45.146237Z  INFO WAL receiver{timeline=57a57583805b132d575748646f24b15f tenant=714c6bb05b524212a57590951cd63bc4}:processing record{lsn=0/16D3308}:get_reconstruct_data{key=000000000000000000000000000000000000 request_lsn=0/16D12B8 reconstruct_state=ValueReconstructState { records: [], img: None }}:search{key=000000000000000000000000000000000000 end_lsn=0/16CAC19}: no layer found for request on 000000000000000000000000000000000000 at 0/16CAC19
```

Actually it does traverse ancestry tree but the key's getting dropped in gc jiggle. Particularly these images look suspicious.  
```
├── [4.0M]  000000067F000032AE000000000000000001-000000067F000032AE00004000000000010B__0000000001696629-0000000001BACBE9
├── [2.7M]  000000067F000032AE00004000000000010B-030000000000000000000000000000000002__0000000001696629-0000000001BACBE9
```
active layers,
```
historic_layers: inmem-0000000001BACBE9-FFFFFFFFFFFFFFFF
historic_layers: 000000067F000032AE000000000000000001-000000067F000032AE00004000000000010B__0000000001696629-0000000001BACBE9
----- delta layer for ten 714c6bb05b524212a57590951cd63bc4 tli 350566f22c2f2033b9f46344822dc1d1 keys 000000067F000032AE000000000000000001-000000067F000032AE00004000000000010B lsn 0/1696629-0/1BACBE9 ----
historic_layers: 000000067F000032AE00004000000000010B-030000000000000000000000000000000002__0000000001696629-0000000001BACBE9
----- delta layer for ten 714c6bb05b524212a57590951cd63bc4 tli 350566f22c2f2033b9f46344822dc1d1 keys 000000067F000032AE00004000000000010B-030000000000000000000000000000000002 lsn 0/1696629-0/1BACBE9 ----
```
disk layout, 
```
test_output/test_branch_and_gc/repo/tenants/714c6bb05b524212a57590951cd63bc4/timelines/350566f22c2f2033b9f46344822dc1d1
├── [5.5M]  000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001BACBE9-0000000001FC35D1
├── [5.5M]  000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001FC35D1-00000000023DFFF1
├── [5.5M]  000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000023DFFF1-00000000027FFFD1
├── [4.0M]  000000067F000032AE000000000000000001-000000067F000032AE00004000000000010B__0000000001696629-0000000001BACBE9
├── [2.7M]  000000067F000032AE00004000000000010B-030000000000000000000000000000000002__0000000001696629-0000000001BACBE9
├── [   0]  ephemeral-7
└── [ 512]  metadata
test_output/test_branch_and_gc/repo/tenants/714c6bb05b524212a57590951cd63bc4/timelines/57a57583805b132d575748646f24b15f
├── [   0]  ephemeral-4
└── [ 512]  metadata
test_output/test_branch_and_gc/repo/tenants/714c6bb05b524212a57590951cd63bc4/timelines/be2627bcc88826cb8591acd042f72263
├── [ 24M]  000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000016965B0-0000000001696629
└── [ 512]  metadata
```

**Comment by hlinnaka** [comment]
What's the plan with this PR, @dhammika? This doesn't reproduce issue 707, because if it did, it wouldn't be passing. Is it adding coverage for some other scenario that's not covered by current tests? Is that visible in the code coverage report?

**Comment by dhammika** [comment]
> What's the plan with this PR, @dhammika? This doesn't reproduce issue 707, because if it did, it wouldn't be passing. Is it adding coverage for some other scenario that's not covered by current tests? Is that visible in the code coverage report?

This was an attempt to create image layout to repro #707 but I couldn't repro it with just twiddling gc and compaction params. L0 image threshold changes surfaced a separate issue  though #1594. IMO original issue is still there even in new redone image layers.

**Comment by aome510** [comment]
@hlinnaka can you review this PR now? I managed to re-produce #707 in https://github.com/neondatabase/neon/tree/thang/exp-07-04 by reverting your fix.

File: test_runner/batch_others/test_branch_and_gc.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import NeonEnv
+from fixtures.utils import lsn_from_hex
+
+
+# Test the GC implementation when running with branching
+# This test reproduces the issue https://github.com/neondatabase/neon/issues/707.
+#
+# Consider two LSNs `lsn1` and `lsn2` with some delta files as folows:
```
Comment (hlinnaka) [inline test_runner/batch_others/test_branch_and_gc.py:9]: typo:
```suggestion
# Consider two LSNs `lsn1` and `lsn2` with some delta files as follows:
```

File: test_runner/batch_others/test_branch_and_gc.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    lsn1 = main_cur.fetchone()[0]
+    log.info(f'LSN1: {lsn1}')
+
+    main_cur.execute('INSERT INTO foo SELECT FROM generate_series(1, 100000)')
+    main_cur.execute('SELECT pg_current_wal_insert_lsn()')
+    lsn2 = main_cur.fetchone()[0]
+    log.info(f'LSN2: {lsn2}')
+
+    # set the GC horizon such that it doesn't cover lsn1 so that
+    # we can create a new branch starting from lsn1
```
Comment (hlinnaka) [inline test_runner/batch_others/test_branch_and_gc.py:83]:
```suggestion
    # Set the GC horizon so that lsn1 is outside the horizon.
```

Comment (hlinnaka) [reply test_runner/batch_others/test_branch_and_gc.py:83]: I'm confused though. If 'lsn1' is outside the GC horizon, creating the branch below will fail, no?

**Review by hlinnaka — APPROVED** [review]
I find the comments hard to understand. And I didn't fully understand the diagram. I made some small suggestions, but maybe try rewriting all the comments from scratch with your own words, that might result in a more coherent story.

Other than that, LGTM. Feel free to push it when you think it's ready.

File: test_runner/batch_others/test_branch_and_gc.py (hunk 3 above)
Comment (aome510) [reply test_runner/batch_others/test_branch_and_gc.py]: Yeah it should be "so that lsn1 is **inside** the horizon"

**Comment by aome510** [comment]
> I find the comments hard to understand. And I didn't fully understand the diagram. I made some small suggestions, but maybe try rewriting all the comments from scratch with your own words, that might result in a more coherent story.

Actually, the whole comment describing the test was written by me. I tried to make it a bit easier to understand now. Hope this is better

**Review by hlinnaka — APPROVED** [review]

**Pull request merged**